com.unity.multiplayer.mlapi/Runtime/Timing/ClientNetworkTimeProvider.cs

```diff
@@ -0,0 +1,121 @@
+        /// </summary>
+        public float TargetClientBufferTime { get; set; }
+
+        /// <summary>
+        /// Gets or sets the amount of the time in seconds the client should buffer incoming messages from the server. This increases server time.
+        /// A higher value increases latency but makes the game look more smooth in bad networking conditions.
+        /// </summary>
+        public float TargetServerBufferTime { get; set; }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="ClientNetworkTimeProvider"/> class with default buffer values.
+        /// </summary>
+        /// <param name="networkStats">The <see cref="INetworkStats"/> to get network information from.</param>
+        /// <param name="tickRate">The tickrate.</param>
+        public ClientNetworkTimeProvider(INetworkStats networkStats, int tickRate)
+        {
+            m_NetworkStats = networkStats;
+
+            // default buffer values. Can be manually overriden.
+            TargetClientBufferTime = 1f / tickRate;
+            TargetServerBufferTime = 2f / tickRate;
+        }
+
+        /// <inheritdoc/>
+        public bool AdvanceTime(ref NetworkTime predictedTime, ref NetworkTime serverTime, float deltaTime)
+        {
+            // advance time
+            predictedTime += deltaTime * PredictedTimeScale;
+            serverTime += deltaTime * ServerTimeScale;
+
+            // time scale adjustment based on whether we are behind / ahead of the server in terms of inputs
+            // This implementation uses RTT to calculate that without server input which is not ideal. In the future we might want to add a field to the protocol which allows the server to send the exact input buffers size to the client.
+            float lastReceivedSnapshotTime = m_NetworkStats.LastReceivedSnapshotTick.Time;
+            float targetServerTime = lastReceivedSnapshotTime - TargetClientBufferTime;
+            float targetPredictedTime = lastReceivedSnapshotTime + m_NetworkStats.Rtt + TargetServerBufferTime;
+
+            // Reset timescale. Will be recalculated based on new values.
+            bool reset = false;
+            PredictedTimeScale = 1f;
+            ServerTimeScale = 1f;
+
+            // reset because too large predicted offset?
+            if (predictedTime.FixedTime < targetPredictedTime - HardResetThreshold || predictedTime.FixedTime > targetPredictedTime + HardResetThreshold)
+            {
+                reset = true;
+            }
+
+            // reset because too large server offset?
+            if (serverTime.FixedTime < targetServerTime - HardResetThreshold || serverTime.FixedTime > targetServerTime + HardResetThreshold)
+            {
+                reset = true;
+            }
+
+            // Always reset both times to not break simulation integrity.
+            if (reset)
+            {
+                predictedTime = new NetworkTime(predictedTime.TickRate, targetPredictedTime);
```

Comment:
I wonder if any other system would want to know when the reset condition happens and if so would we want a delegate to fire here?  Or even for testing if you want to force the condition and verify it happened

Reply:
That's why the INetworkTimeProvider returns true/false based on whether there is a reset. This value is currently not propagated to other systems internally but our interface allows to do this without breaking backwards compatibility later.